File: leetcode/350.py
"""
Runtime: 44 ms, faster than 82.93% of Python3 online submissions for Intersection of Two Arrays II.
Memory Usage: 14.5 MB, less than 38.31% of Python3 online submissions for Intersection of Two Arrays II.
"""
from collections import Counter


class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # nums1 is the longer of the two lists
        if len(nums1) < len(nums2):
            return self.intersect(nums2, nums1)

        # hashmaps
        cnt1, cnt2 = Counter(nums1), Counter(nums2)

        # iterate through the shorter list
        res = []
        for n in nums2:
            if n in cnt1.keys() and n in cnt2.keys() and cnt1[n] and cnt2[n]:
                cnt1[n] -= 1
                cnt2[n] -= 1
                res.append(n)

        return res

File: leetcode/350.cpp
// Runtime: 3 ms
// Memory Usage: 11.5 MB
/*
  O(n + m) create the count maps.
  Use the unordered_map for O(1) avg lookup time.
  O(min(m, n)) to add intersect elements.
*/
class Solution {
 public:
  vector<int> intersect(vector<int>& nums1, vector<int>& nums2) {
    // iterate over the shorter
    if (nums1.size() > nums2.size()) return intersect(nums2, nums1);

    // count n1, n2 occurrences
    std::unordered_map<int, int> n1_count, n2_count;
    for (auto n : nums1) n1_count[n] += 1;
    for (auto n : nums2) n2_count[n] += 1;

    // fill the intersection with min count of intersected elements
    vector<int> res = {};
    for (auto pair : n1_count) {
      int count = std::min(n1_count[pair.first], n2_count[pair.first]);
      for (int i = 0; i < count; ++i) res.push_back(pair.first);
    }

    return res;
  }
};
